### ISSUE TITLE
Typescript projects: outDir in tsconfig.json forces always rebuild of project

### ISSUE THREAD

**Issue by jeffrson** [Microsoft/nodejstools#1793]
##### Expected Behavior 
Even when "outDir" is used in tsconfig.json, a project shouldn't be rebuilt if there's no change to the source files.

##### Actual Behavior 
.js files in "outDir" (different from a certain srcDir) are not recognized when VS decides whether a project should be rebuild

- NTVS Version: 1.4.10918.1
- Visual Studio Version: 15.4.4
- Node.js Version: 9.2.0

##### Steps to Reproduce
1. create a new NodeJS Typescript project
2. build it, run it; run it again and again - no rebuild occurrs
3. create an empty folder, e.g. "out" and put >"outDir": "./out"< in tsconfig.json compilerOptions
4. build it, run it - run it again and again - every  time VS asks whether it should rebuild the project

**Comment by paulvanbrenk** [comment]
Thanks for reporting, this is an issue in the TypeScript tools for VS, which should be fixed in a future update of VS 2017. (Probably update 6)

**Comment by jeffrson** [comment]
Is there a specific issue I could follow? It's not fixed in Update 6, so I would like to get information about the actual progress happening.

**Comment by uniqueiniquity** [comment]
@jeffrson I apologize for the confusion.
The issue has been addressed, but it depends on the version of TypeScript you are targeting (i.e., it was fixed in TypeScript 2.6). In order to get the fix, you will have to retarget your project for TS 2.6 or higher.

**Comment by jeffrson** [comment]
@uniqueiniquity thanks for stepping in.
Just to be sure - targetting TS means changing it in VS project properties combobox, right?
I tried with 2.6, 2.7 and "latest available", with (MSBuild) project settings or tsconfig.json. Everytime I run a simple TS (console.log) VS prompts me because the project should be outdated. So actually nothing changed.
Anything else I could try?

**Comment by uniqueiniquity** [comment]
Yes, that is what I meant.
I'm not sure I know what you mean by "VS prompts me because the project should be outdated". Could you post a screenshot?

**Comment by uniqueiniquity** [comment]
Oh I think I may see what you mean. Let me look into this.

**Comment by uniqueiniquity** [comment]
Seems like this is an NTVS issue after all. We will reopen and look into this. Thank you for your feedback!

**Comment by uniqueiniquity** [comment]
Seems like this is an NTVS issue after all. We will reopen and look into this. Thank you for your feedback!

**Comment by JMS-1** [comment]
I didn't find the time to retest myself but if the bug is still there perhaps you can take a look on my thoughts in #1794 (linked above) and the MSBuild diagnostics log.

**Comment by paulvanbrenk** [comment]
Found one issue, don't have a fix (yet)…. we consider the contents of the node_modules folder to determine if we need to rebuild, since those files are never updated we rebuild every time.

There are probably other issues....

**Comment by jeffrson** [comment]
The issue reported is completely independent from node_modules.

Say we have this `app.ts`:
``` Typescript
console.log('Hello world')
```
:-) no dependencies (posted for completness)
``` json
{
  "name": "tsc_outdir_test",
  "version": "0.0.0",
  "description": "tsc_outdir_test",
  "main": "app.ts",
  "dependencies": {}
}
```

and this compiler config (tsconfig.json):
``` json
{
  "compilerOptions": {
    "outDir": "build"
  },
  "exclude": [
    "node_modules"
  ]
}
```

If F5 (start debug) is pressed, the project is reported as being outdated - every time.
As soon as "outDir" is removed, the project is rebuilt only once. And, btw, if I manage to copy output files from `build` back to where app.ts resides, then there's also no rebuilt.

The issue must be related to failure of localizing app.js, app.js.map.

**Comment by paulvanbrenk** [comment]
This looks like it's much more work than I expected. Closing in favor for #1949 which captures more use cases.
